runtime/runtime/src/lib.rs

```diff
@@ -155,6 +155,79 @@ pub struct ApplyState {
     pub bandwidth_requests: BlockBandwidthRequests,
 }
 
+/// Represents a batch of transactions sharing the same (signer_id, public_key),
+/// sorted by ascending nonce.
+struct TransactionBatch<'a> {
+    pub indices: &'a [usize],
+    pub signed_txs: &'a [SignedTransaction],
+}
+
+impl<'a> TransactionBatch<'a> {
+    fn iter(&self) -> impl Iterator<Item = (usize, &SignedTransaction)> {
+        self.indices.iter().map(|idx| (*idx, &self.signed_txs[*idx]))
```

Comment:
i think you should add an assert that the lengths are the same.

Should the function consume `self`?

```suggestion
        self.indices.into_iter().map(|idx| (*idx, &self.signed_txs[*idx]))
```
Wouldn't this work too?